firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.core;
+
+import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.util.Function;
+import com.google.firestore.v1.StructuredQuery.CompositeFilter.Operator;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Represents a filter that is the conjunction or disjunction of other filters. */
+public class CompositeFilter extends Filter {
+  private final List<Filter> filters;
+  private final Operator operator;
+
+  public CompositeFilter(List<Filter> filters, Operator operator) {
+    this.filters = filters;
+    this.operator = operator;
+  }
+
+  public List<Filter> getFilters() {
+    return filters;
+  }
+
+  public Operator getOperator() {
+    return operator;
+  }
+
+  @Override
+  public List<FieldFilter> getAllFieldFilters() {
+    List<FieldFilter> result = new ArrayList<>();
+    for (Filter subfilter : filters) {
+      result.addAll(subfilter.getAllFieldFilters());
+    }
+    return result;
+  }
+
+  /**
+   * Returns the first inequality filter contained within this composite filter. Returns null if it
+   * does not contain any inequalities.
+   */
+  @Override
+  public FieldPath inequalityField() {
+    FieldFilter found = firstFieldFilterWhere(f -> f.isInequality());
+    if (found != null) {
+      return found.getField();
+    }
+    return null;
+  }
+
+  public boolean isConjunction() {
+    return operator == Operator.AND;
+  }
+
+  public boolean isDisjunction() {
+    // TODO(orquery): Replace with Operator.OR.
+    return operator == Operator.OPERATOR_UNSPECIFIED;
+  }
+
+  /**
+   * Performs a depth-first search to find and return the first FieldFilter in the composite filter
+   * that satisfies the condition. Returns null if none of the FieldFilters satisfy the condition.
+   */
+  private FieldFilter firstFieldFilterWhere(Function<FieldFilter, Boolean> condition) {
```

Comment:
```suggestion
  @Nullable private FieldFilter firstFieldFilterWhere(Function<FieldFilter, Boolean> condition) {
```

Reply:
Also, just because I can't help myself and to give you something to ignore: What about "findFirstMatchingFilter"? This method name could use a verb.